### ISSUE TITLE
exposure.equalize_hist doesn't output the same type as the input

### ISSUE THREAD

**Issue by TheShadow29** [scikit-image/scikit-image#2677]
I am not completely sure if it should be a feature request or a bug.
Currently exposure.equalize_hist given an image of input of type uint8 outputs an image of type float64. The proposal is to give the output in uint8 format itself. 
Current implementation:
```
img1 = io.imread(path_to_file)
img11 = skimage.exposure.equalize_hist(img1)
print('Original Image is of dtype ', img1.dtype)
print('Histogram Equalized Image is of dtype', img11.dtype)
```
Output:
```
('Original Image is of dtype ', dtype('uint8'))
('Histogram Equalized Image is of dtype', dtype('float64'))
```
#1192 mentions this, but the issue is not really addressed.
To be honest I am not even sure why the output image is of type 'float64'. At no point of time should  image of type uint8 should be changed to float64, i.e. there should be no case where we should be working with floats, we should be in integer domain itself. Here is a nice text describing histogram equalization https://www.math.uci.edu/icamp/courses/math77c/demos/hist_eq.pdf

Histogram equalization would simply be a transformation curve, mapping one intensity to another corresponding intensity (the formulae given in the mentioned link) and it should be all in the integer domain when the input is of type uint8. 

Also as far as I know, MATLAB also does the same thing, that is given an input of uint8 returns output in uint8 itself. Link for the same https://www.mathworks.com/help/images/ref/histeq.html

In short, the current implementation of the equalize_hist is doing something similar but will give a different result from what a normal histogram equalization would return at least for an uint8 image. Also I think it is more sensible to return image of the same dtype.

**Comment by TheShadow29** [comment]
@soupault can you have a look at this too?

**Comment by jni** [comment]
Hi @TheShadow29 

I think the difference is much smaller than you think. One of the perils of working with uint8 instead of float! `abs(img12 - img2)` will result in negative values, anywhere `img12` is smaller than `img2`, looping back to 255. Try `diff = abs(img12.astype(float) - img2.astype(float)).astype('uint8')`, then compute the mean and max of that image.

I don't know that it's generally true that "it should be all in the integer domain when the input is of type uint8". The choice of `floor` in the paper you linked is a consequence of the desire to work in the integer domain. If you work in float, you don't need to use the floor at all, but rather you can use the exact value from the mapping, and that will be a more faithful histeq result.

Could you post links to your rice image and the Matlab-equalised image? I can investigate a bit more if I have some sample test data.

**Comment by TheShadow29** [comment]
@jni I should've been more careful in using the absolute value it seems. Here are the images you requested.
![rice](https://user-images.githubusercontent.com/10602903/27295614-5d721c18-54d2-11e7-935d-908fe24bb96d.png)
![rice_eq](https://user-images.githubusercontent.com/10602903/27295634-7290edae-54d2-11e7-9dcf-81ae05770daa.png)

**Comment by TheShadow29** [comment]
> I don't know that it's generally true that "it should be all in the integer domain when the input is of type uint8". The choice of floor in the paper you linked is a consequence of the desire to work in the integer domain. If you work in float, you don't need to use the floor at all, but rather you can use the exact value from the mapping, and that will be a more faithful histeq result.

Yes its true that it would be a more faithful histeq. But I feel that it shouldn't really be the case when the input image is of uint8. I  mean when the input is itself has 8bit precision it seems counter intuitive to use more precision in the intermediate process. 

Anyways here is the new difference image. And yes now it does seem like the error is much lower than I expected. I had to get rid of the `io.use_plugin('matplotlib')` line though. For some reason the img2 was being shown in float.
![git_post_skimage_3](https://user-images.githubusercontent.com/10602903/27296302-2052651a-54d5-11e7-9a7e-b0f94f0b6722.png)

**Comment by TheShadow29** [comment]
Excellent description and comparison. Thanks a lot @jni
